sensor/common/clusterentities/store_ips.go

```diff
@@ -0,0 +1,244 @@
+	e.updateMetricsNoLock()
+}
+
+func (e *podIPsStore) historyEnabled() bool {
+	return e.memorySize > 0
+}
+
+func (e *podIPsStore) updateMetricsNoLock() {
+	metrics.UpdateNumberOfIPs(len(e.ipMap), len(e.historicalIPs))
+}
+
+func (e *podIPsStore) RecordTick() bool {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	removed := false
+	for ip, m := range e.historicalIPs {
+		for deploymentID, status := range m {
+			status.recordTick()
+			// Remove all historical entries that expired in this tick.
+			removed = e.removeFromHistoryIfExpired(deploymentID, ip) || removed
+		}
+	}
+	return removed
+}
+
+func (e *podIPsStore) Apply(updates map[string]*EntityData, incremental bool) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.applyNoLock(updates, incremental)
+}
+
+func (e *podIPsStore) applyNoLock(updates map[string]*EntityData, incremental bool) {
+	defer e.updateMetricsNoLock()
+	if !incremental {
+		for deploymentID := range updates {
+			e.purgeDeploymentNoLock(deploymentID)
+		}
+	}
+	for deploymentID, data := range updates {
+		if data.isDeleteOnly() {
+			// A call to Apply() with empty payload of the updates map (no values) is meant to be a delete operation.
+			continue
+		}
+		e.applySingleNoLock(deploymentID, *data)
+	}
+}
+
+func (e *podIPsStore) purgeDeploymentNoLock(deploymentID string) {
+	if e.historyEnabled() {
+		e.moveDeploymentToHistory(deploymentID)
+	} else {
+		e.deleteDeploymentFromCurrent(deploymentID)
+		// In case we allow in the future to disable history during runtime, we would need to remove here all
+		// expired data for deploymentID from the history.
+	}
+}
+
+func (e *podIPsStore) LookupByNetAddr(ip net.IPAddress, port uint16) (results, historical []LookupResult) {
+	e.mutex.RLock()
+	defer e.mutex.RUnlock()
+	for deploymentID := range e.ipMap[ip] {
+		result := LookupResult{
+			Entity:         networkgraph.EntityForDeployment(deploymentID),
+			ContainerPorts: []uint16{port},
+		}
+		results = append(results, result)
+	}
+	// if there is a match in the map, then there is no need to search in history,
+	// as it may contain data about different past deployment using this address
+	for histDeploymentID := range e.historicalIPs[ip] {
+		result := LookupResult{
+			Entity:         networkgraph.EntityForDeployment(histDeploymentID),
+			ContainerPorts: []uint16{port},
+		}
+		historical = append(historical, result)
+	}
+	return results, historical
+}
+
+func (e *podIPsStore) applySingleNoLock(deploymentID string, data EntityData) {
+	ipsSet := e.reverseIPMap[deploymentID].Unfreeze()
+	for ip := range data.ips {
+		ipsSet.Add(ip)
+
+		// check if this IP already belongs to other deployment
+		deplSet := e.ipMap[ip]
+		deplSet.Add(deploymentID)
+		// This IP has more than one deployment! Interesting, let's record it.
+		if deplSet.Cardinality() > 1 {
+			metrics.ObserveManyDeploymentsSharingSingleIP(ip.AsNetIP().String(), deplSet.AsSlice())
+		}
+		e.ipMap[ip] = deplSet
```

Comment:
Maybe it's just me, but I find this is a little bit difficult to read. This works because `deplSet := e.ipMap[ip]` will create an empty set if the `ip` is not in the `ipMap`. If instead of a map of sets we had a map of maps this wouldn't work. I'm not sure if I'd prefer to change this to something like:

```golang
depSet, found := e.ipMap[ip]
if !found {
    depSet = NewStringSet(deploymentID)
}
[...]
```
Or simply add a comment explaining why this works, `if ip is not in the ipMap then e.ipMap[ip] returns an empty StringSet`.

Reply:
Yes, I use exactly that trick here. The code with the map instead of set was one-tick more complex due to that. I will add the comment.